src/NHibernate/Collection/Generic/PersistentGenericMap.cs

```diff
@@ -568,5 +578,95 @@ public void Operate()
 		}
 
 		#endregion
+
+		[Serializable]
+		private class ValuesWrapper : ICollection<TValue>, IQueryable<TValue>
+		{
+			private readonly PersistentGenericMap<TKey, TValue> _map;
+
+			public ValuesWrapper(PersistentGenericMap<TKey, TValue> map)
+			{
+				_map = map;
+			}
+
+			#region IQueryable<TValue> Members
+
+			[NonSerialized]
+			private IQueryable<TValue> _queryable;
+
+			Expression IQueryable.Expression => InnerQueryable.Expression;
+
+			System.Type IQueryable.ElementType => InnerQueryable.ElementType;
+
+			IQueryProvider IQueryable.Provider => InnerQueryable.Provider;
+
+			private IQueryable<TValue> InnerQueryable => _queryable ?? (_queryable = new NhQueryable<TValue>(_map.Session, _map));
+
+			#endregion
+
+			#region ICollection<TValue> Members
+
+			public IEnumerator<TValue> GetEnumerator()
+			{
+				_map.Read();
+				return _map.WrappedMap.Values.GetEnumerator();
+			}
+
+			IEnumerator IEnumerable.GetEnumerator()
+			{
+				_map.Read();
+				return GetEnumerator();
+			}
+
+			public void Add(TValue item)
```

Comment:
Dictionary's "Values" collection is read-only.

Reply:
Must still implement the interface, and I am just letting the throw be done by the underlying collection. But just for doing that, it triggers the read. Would you rather have a direct throw not reading at all? Since it is an error case anyway, I do not tend to optimize it.

Reply:
I don't think this actually matters much, so let it be this way.

Reply:
It does not matter much but I have already changed it anyway, just not yet pushed.

Reply:
But how about "I do not tend to optimize it."?